### PR TITLE
[SPMD] Support manual all-reduce

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This is to add manual all-reduce support to SPMD and it currently only supports one input tensor. For array support, we can do that in python layer instead.

Test Plan:
python ./test/spmd/test_xla_sharding.py -v -k test_spmd_all_reduce

### CODE REVIEW

File: torch_xla/csrc/tensor_methods.cpp [hunk 1]
```diff
@@ -392,6 +392,13 @@ void all_reduce(const std::vector<XLATensorPtr>& inputs,
   }
 }
 
+XLATensorPtr all_reduce(const XLATensorPtr& input, AllReduceType reduce_type,
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_methods.cpp:395]: can you call it `all_reduce _no_token`, the only difference in signature is it does not take `pin_layout` but the main difference in the op is that it does not set token.. It is better to reflect that in the name.

**Review by JackCaoG — APPROVED** [review]
approve to unblock, but I think we should fix the tensor method name

**Comment by JackCaoG** [comment]
for array support  do you plan to call `all_reduce` multiple times? In our C++ implementation I think we group tensors by dtype and call all_rduce once per dtype.

File: torch_xla/csrc/tensor_methods.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch_xla/csrc/tensor_methods.cpp:395]: Sure. I can follow up with that.

**Comment by alanwaketan** [comment]
> for array support do you plan to call `all_reduce` multiple times? In our C++ implementation I think we group tensors by dtype and call all_rduce once per dtype.

I don't think that's necessary. I'm thinking the compiler should be smart enough to fuse all-reduces if the fusion is necessary.

**Comment by alanwaketan** [comment]
Thanks Jack for approving.

**Pull request merged**